corfudb-tools/src/main/java/org/corfudb/compactor/CompactorCheckpointer.java

```diff
@@ -77,8 +82,10 @@ protected void startCheckpointing() {
             log.error("Sleep interrupted with exception: ", ie);
         } catch (Exception e) {
             log.error("Exception during checkpointing: {}, StackTrace: {}", e.getMessage(), e.getStackTrace());
-        } finally {
-            distributedCheckpointer.shutdown();
         }
     }
+
+    public void shutdown() {
```

Comment:
I think in general it's a good rule to write the shutdown() method of class is to clean up only the resources the class creates. Since cpRuntime and corfuRuntime (L29) are both initialized by this class, I think we should just shutdown them here rather than in ServerTriggeredCheckpointer. This makes it easier to track where the resources are cleaned up - think about that the runtime is passed to both ServerTriggeredCheckpointer and DistributedCheckpointer, how do I know which class handles the shutdown of the runtime?

Reply:
+1 - I think it becomes much easier to manage as well.

Right now, imagine a case where `cpRuntime` succeeds in connecting but not `corfuRuntime`. Then the allocation of `corfuCompactorMain` does not finish and remains null, so we are still leaking resources.